71_RayTracingPipeline/app_resources/raytrace.rgen.hlsl

```diff
@@ -41,22 +52,58 @@ void main()
         const float32_t4 tmp = mul(pc.invMVP, float32_t4(d.x, d.y, 1, 1));
         const float32_t3 targetPos = tmp.xyz / tmp.w;
 
-        float32_t3 direction = normalize(targetPos - pc.camPos);
+        const float32_t3 camDirection = normalize(targetPos - pc.camPos);
 
         RayDesc rayDesc;
         rayDesc.Origin = pc.camPos;
-        rayDesc.Direction = direction;
-        rayDesc.TMin = 0.001;
+        rayDesc.Direction = camDirection;
+        rayDesc.TMin = 0.01;
         rayDesc.TMax = 10000.0;
         
-        ColorPayload payload;
-        payload.seed = seed;
+        PrimaryPayload payload;
+        payload.alphaThreshold = nextRandomUnorm(rnd);
         TraceRay(topLevelAS, RAY_FLAG_NONE, 0xff, ERT_PRIMARY, 0, EMT_PRIMARY, rayDesc, payload);
 
-        hitValues += payload.hitValue;
+        if (payload.rayDistance < 0)
+        {
+            hitValues += s_clearColor;
+            continue;
+        }
+
+        const float32_t3 worldPosition = pc.camPos + (camDirection * payload.rayDistance);
+        const float32_t3 worldNormal = payload.worldNormal;
+        const Material material = unpackMaterial(payload.material);
```

Comment:
fun fact, passing the geometry & instance ID out with the payload, then loading all the stuff here, puts less pressure on the stack and will probably perform better

Reply:
hmm because the prodecural intersection shaders compute normal and position https://github.com/Devsh-Graphics-Programming/Nabla-Examples-and-Tests/pull/173#discussion_r1929306333, we may have to keep just that in the payload, but the whole material (instead of handle) is a mistake